Support new `.find()` on arrays and `.fullIdxType` on domains/arrays

bradcray opened the pull request:
Here, I'm adding the new array.find() signatures agreed upon in #21423, to replace the old tuple-returning version of array.find().  The version that takes an index by ref and returns a bool is immediately available;  the one-argument version that returns an OOB index for rectangular arrays with integral indices is only available when compiling with `-suseNewArrayFind` since it has the same signature as the old one.

Since these routines take and return the "natural, complete" index type of the array in question (e.g., `int` for a 1D array, but `k*int` for a kD array), I also implemented `fullIdxType` on domains and arrays to represent this type, as discussed in #19141.  As part of this, I updated the documentation for both idxType and fullIdxType to try and clarify their definitions and differences.

I made use of the new interface in a new revcomp study variation, reusing my longstanding `revcomp-blc.chpl` file now that the previous version has become `revcomp5.chpl` on the official site and in the `shootout/submitted/` directory.  This resulted in a clean implementation of the benchmark, though I haven't done anything with the array.find() implementation at this point to try and optimize it.  I also extended my `revcomp-blc-gcc8-algs.chpl` test to include an `array.find()`-based option.

I added some tests exercising the new interface in `test/arrays/methods/find-irreg-onearg.chpl`, `test/arrays/methods/find-irreg.chpl`, and `test/arrays/methods/find.chpl`.

I also updated the `test/arrays/userAPI` and `test/domains/userAPI` tests to exercise `.fullIdxType` and `.find()` (in the former case).

I also re-instated old uses of `find()` that had been removed from module and test code when it was deprecated, to use the new stable interfaces—either the two-arg version in code that we don't control (like `modules/dists/SparseBlockDist.chpl`), or the one-arg version using compopts files (in tests like `test/arrays/diten/arrayFind.chpl`).

I split the array.find() case that was in `test/unstable/arrayFuncs.chpl` into a new test in `test/deprecated/arrayFind.chpl`

Resolves #21423.
Resolves #19141.
Moves us slightly closer to being able to call  #18089 complete.

TODO:
- [x] re-instate some of the uses of `find()` that went away when it was deprecated?  (if we have control over the compiler options anyway...)

bradcray commented on modules/internal/ChapelArray.chpl:
```diff
@@ -1733,15 +1736,56 @@ module ChapelArray {
+
+      Support is currently limited to rectangular arrays.
+    */
+    proc find(val: this.eltType) /* TODO: return type is hard to express */ {
```
@e-kayrakli : With respect to your comment yesterday:

> [when wanting the most productive type for a given array/domain] we ask the user to write the conditional.

This is where I felt that having a way to say `this.mostConvenientIdxType` (or redefining `this.idxType` to do the same thing) would be more productive than writing the conditional and having it show up in the docs.  Basically, my hope would be that we'd have a compact and standard way to write that type rather than having each such routine have a conditional or rely on its own personal helper function (which was the first way I tackled this before realizing that it made the documentation wonky / require too much indirection).

e-kayrakli replied:
When thinking about how to respond to the design issue for array.find, I found myself needing the same thing without looking at this first :)

I am still not quite happy about providing that new thing because (1) I don't think we can find a name that's less then 3 words, which is too long for the array/domain interface, (2) array interface will have 3 fields (or field-like things) where you can always use any 2 to compute the other 3rd. Which is OK, I guess, but a bit of an overkill and somewhat tautological.

Not that I was vehemently against it, but I'm OK with adding something new, despite my reservations above.

bradcray replied:
@e-kayrakli : I'm not sure I'm interpreting your comment correctly.  Are you saying "OK, I agree we should have this even though naming it will be hard" or "I still don't think we should have this, but I can see why you'd want it?"  If the latter, practically speaking, how would you document this routine's return type in the array library documentation today?

e-kayrakli replied:
Definitely the former. Sorry for not being clear.

bradcray replied:
My best ideas for today:
* `accessType`?  `fullIdxType`?
* Or make the current index type `dimIdxType` and the full thing be `idxType`?

e-kayrakli replied:
Note that Andy has started the discussion on this here: https://github.com/chapel-lang/chapel/issues/19141#issuecomment-1406818593 (Just to avoid getting your ideas lost here)

bradcray replied:
I ended up implementing `fullIdxType` as part of this PR.

bradcray commented:
@stonea : I think that this is ready for review, and that you'd be a natural reviewer as a chief contributor to the three issues that spawned it.  Would you have time to do that?

stonea commented on test/deprecated/arrayFind.chpl:
```diff
@@ -0,0 +1,6 @@
+// Marked unstable while we figure out if we want to change the interface to
```
Is this comment right? Is it that the new interface for `find` still unsettled or is that we've settled on a new interface but are using the `useNewArrayFind` `config param` until we can get rid of the old interface after going through the normal deprecation cycle.

bradcray replied:
I never read comments.  :D  I've updated this from the original copy-pasted comment.

stonea replied:
> I never read comments. :D I've updated this from the original copy-pasted comment.

Which is why clear code is always better than good comments: comments get outdated, don't get locked down with tests, and are a violation of the whole "don't repeat yourself" principle of software engineering; though I can't bring myself to say they're worthless /rant.

stonea commented on modules/internal/ChapelArray.chpl:
```diff
@@ -1733,15 +1744,80 @@ module ChapelArray {
     }
 
+
+    /*
```
I'd add a sentence indicating what happens if `val` appears in the array multiple times. I'm assuming it returns the first one according to the iteration order of the domain right?

stonea commented on modules/internal/ChapelArray.chpl:
```diff
@@ -1733,15 +1744,80 @@ module ChapelArray {
+          }
+        }
+      } else {
+        for i in this.domain {
```
Do we want to use a parallel loop or reduction for this? I guess the idea is we'll come back and optimize this later at some point now that we've settled on the interface.

bradcray replied:
Yeah, I didn't want to change the implementation from what we've traditionally used in this PR, though I would like to once we get some performance tests written.  I think a challenge is determining what implementation to use when.  For example, for a find on a short 1D uint(8) array, memchr() may be best.  For a long one, some parallel application of memchr() probably would be.  For a short array of 64-bit values, sequential search may be best; for a big one, a maxloc reduction probably would be.  Some domain maps may want (or need) to provide their own implementations for efficiency.  I worry a bit about how many implementations we'll end up wanting and how we'll choose between them, so didn't want to get into that now.

stonea commented on modules/internal/ChapelArray.chpl:
```diff
@@ -1733,15 +1744,80 @@ module ChapelArray {
+      will be returned.  If ``val`` is not found,
+      ``domain.lowBound-1`` will be returned instead.
+
+      Note that for arrays with ``idxType=int(?w)`` (signed ``int``
```
This would be an issue for arrays with unsigned integers too if the low bound was 0 right? I think I'd just generalize the statement to say if the low bound in a dimension is min(idxType) the result is not well-defined.

bradcray replied:
I don't think so.  My understanding is that unsigned integer wraparound is well-defined (so the result will be `max(uint(w))`.  Rather than spelling that out explicitly, I was hoping we could rely on the user understanding (or figuring out) that `0-1` would be `max(uint(w))` for a uint idxType.

stonea reviewed (approved):
Looks good, I just have some minor suggestions about doc. Thanks for resurrecting `find()` in all the applicable tests.

bradcray commented on modules/internal/ChapelArray.chpl:
```diff
@@ -255,6 +255,9 @@ module ChapelArray {
   pragma "no doc"
   config param capturedIteratorLowBound = defaultLowBound;
 
+  config param useNewArrayFind = false;
```
Good catch, I've just pushed a doc string for this config.

bradcray commented:
@stonea : Thanks for the review!  I think I've addressed your comments, and also did a minor wordsmithing + formatting improvement to the deprecation message.  Let me know whether you want to re-review or look at the latest commits before proceeding.

stonea commented:
> Let me know whether you want to re-review or look at the latest commits before proceeding

I'm good with shipping it; thanks for addressing my comments.
